fdbclient/ThreadSafeTransaction.cpp

```diff
@@ -284,6 +285,107 @@ ThreadFuture<Standalone<VectorRef<const char*>>> ThreadSafeTransaction::getAddre
+	}
+
+	Standalone<VectorRef<BlobGranuleChunkRef>> files = getFilesFuture.get();
+
+	try {
+		RangeResult results;
+		// FIXME: could submit multiple chunks to start_load_f in parallel?
+		for (BlobGranuleChunkRef& chunk : files) {
+			RangeResult chunkRows;
+
+			int64_t snapshotLoadId;
+			int64_t deltaLoadIds[chunk.deltaFiles.size()];
+
+			// FIXME: move to transactions?
+			if (doMaterialize) {
+				// Start load process for all files in chunk
+				// In V1 of api snapshot is required, optional is just for forward compatibility
+				ASSERT(chunk.snapshotFile.present());
+				std::string snapshotFname = chunk.snapshotFile.get().filename.toString();
+				snapshotLoadId = granule_context.start_load_f(snapshotFname.c_str(),
+				                                              snapshotFname.size(),
+				                                              chunk.snapshotFile.get().offset,
+				                                              chunk.snapshotFile.get().length,
+				                                              granule_context.userContext);
+				int64_t deltaLoadIds[chunk.deltaFiles.size()];
+				int64_t deltaLoadLengths[chunk.deltaFiles.size()];
+				StringRef deltaData[chunk.deltaFiles.size()];
+				for (int deltaFileIdx = 0; deltaFileIdx < chunk.deltaFiles.size(); deltaFileIdx++) {
+					std::string deltaFName = chunk.deltaFiles[deltaFileIdx].filename.toString();
+					deltaLoadIds[deltaFileIdx] = granule_context.start_load_f(deltaFName.c_str(),
+					                                                          deltaFName.size(),
+					                                                          chunk.deltaFiles[deltaFileIdx].offset,
+					                                                          chunk.deltaFiles[deltaFileIdx].length,
+					                                                          granule_context.userContext);
+					deltaLoadLengths[deltaFileIdx] = chunk.deltaFiles[deltaFileIdx].length;
+				}
+
+				// once all loads kicked off, load data for chunk
+				StringRef snapshotData(granule_context.get_load_f(snapshotLoadId, granule_context.userContext),
+				                       chunk.snapshotFile.get().length);
+				for (int i = 0; i < chunk.deltaFiles.size(); i++) {
+					deltaData[i] = StringRef(granule_context.get_load_f(deltaLoadIds[i], granule_context.userContext),
+					                         chunk.deltaFiles[i].length);
+				}
+
+				// materialize rows from chunk
+				chunkRows = materializeBlobGranule(chunk, keyRange, readVersionOut, snapshotData, deltaData);
+			}
+
+			results.arena().dependsOn(chunkRows.arena());
+			results.append(results.arena(), chunkRows.begin(), chunkRows.size());
+
+			if (doMaterialize) {
+				granule_context.free_load_f(snapshotLoadId, granule_context.userContext);
+				for (int i = 0; i < chunk.deltaFiles.size(); i++) {
+					granule_context.free_load_f(deltaLoadIds[i], granule_context.userContext);
+				}
+			}
+		}
+		return results;
```

Comment:
Historically, the thread safe transaction layer is meant to be a pretty thin wrapper over lower level APIs. I don't know if it's possible or not, but if any of this could be pushed down a layer that would probably be preferred.

Reply:
I think the reason it's implemented here is to perform the cpu-intensive work off the network thread. Maybe we need a comment explaining that (if that's right)

Reply:
Yeah I think you are correct, Josh and I are having a side conversation about this.